trunner/test/harnesses/test_micropython.py

```diff
@@ -0,0 +1,503 @@
+        exp_outputs = [
+            'abc[1-2]xyz',
+            'GHJKL?\\?\\{}}345',
+            'qwer.poiuyy*'
+        ]
+
+        assert exp_outputs == get_outputs(inputs)
+
+    def test_create_regex(self):
+
+        def create_and_check_regex(text_for_regex, matching_texts):
+            regex = create_regex(text_for_regex)
+            for text in matching_texts:
+                print(f"{regex}\n\n{text}")
+                assert re.fullmatch(regex, text) is not None
+
+        # Testing checking regex for a number
+        input = 'abc\\\\d\\+xyz'
+        matching_texts = [
+            'abc1xyz',
+            'abc12xyz',
+            'abc123xyz',
+            'abc12345678912345678xyz',
+            'abc9999999xyz'
+        ]
+
+        create_and_check_regex(input, matching_texts)
+
+        # Testing new line handling
+        input = 'abc\r\r\r\nxyz'
+        matching_texts = [
+            'abc\r\nxyz',
+            'abc\r\r\nxyz',
+            'abc\r\r\r\nxyz',
+            'abc\r\r\r\r\nxyz'
+        ]
+
+        create_and_check_regex(input, matching_texts)
+
+        # Testing handling ########
+        # In MicroPython tests ######## matches zero or more lines with text
+        input = '########'
+        matching_texts = [
+            '\n',
+            'Elizabeth have a cat\n',
+            'Hello world\nHello people\n',
+            '''Lorem ipsum dolor sit amet,
+             consectetur adipiscing elit,
+             sed do eiusmod tempor incididunt ut
+             labore et dolore magna aliqua.
+             '''
+        ]
+
+        create_and_check_regex(input, matching_texts)
+
+        # Testing sequence which can lead to incorrect regex
+        # Some tests uses such a syntax
+        input = '########\r\r\n\\.\\+00'
+        matching_texts = [
+            '00',
+            'Hello world - 00',
+            '''Lorem ipsum dolor sit amet,
+            consectetur adipiscing elit,
+            sed do eiusmod tempor incididunt ut
+            labore et dolore magna aliqua. 00'''
+        ]
+
+        create_and_check_regex(input, matching_texts)
+
+        # Testing new lines around ########
+        input = 'start\r\n########\r\nstop'
+        matching_texts = [
+            '''start\r\nstop''',
+
+            '''start\r\nline1\r\nstop''',
+
+            '''start\r\nline1\r\nline2\r\nstop''',
+
+            '''start\r\nline1\r\r\nline2\r\r\r\nstop'''
+        ]
+
+        create_and_check_regex(input, matching_texts)
+
+
+class TestStandardHarness:
+
+    # Program for communicating with harness
+    def fake_program(socket, args):
+        test_path = args[0]
+        program_output = args[1]
+        exp_output = args[2]
+
+        send_through_socket(socket, program_output)
+        # b = program_output.encode("utf-8")
+        # socket.send(b)
+
+        if exp_output is not None:
+            recv_exp_through_socket(socket, f"cat {test_path}.exp\n")
+            send_through_socket(socket, exp_output)
```

Comment:
```suggestion
    def harness_callback(mock, *args):
        program_output = args[0]
        mock.send(program_output)

        if len(args) > 1:
            test_path = args[1]
            exp_output = args[2]
            mock.recv_exp(f"cat {test_path}.exp\n")
            mock.send(exp_output)
```
suggestion how it can looks with "mock" object that I mentioned before